### DISCUSSION TITLE
How to stretch image into A4 PDF?

### DISCUSSION THREAD

**Discussion by feklee** [ImageMagick/ImageMagick#2478] [General] [answered]
*Input:* JPG image with *different aspect ratio* than A4.

*Desired output:* PDF that contains the JPG image stretched so that it fills the page and so that it has exactly A4 aspect ratio. Ideally, the JPEG image is not resampled, just stretched using PDF commands.

*Can ImageMagick do this? If not, what other command line tools would be a good option?*

**Comment by fmw42** [comment]
Imagemagick is a raster processor. So any resizing would resample or repeat pixels. When reading a PDF, ImageMagick uses Ghostscript, which will rasterize your PDF.

**Comment by feklee** [comment]
> When reading a PDF, ImageMagick uses Ghostscript, which will rasterize your PDF.

Read my question again. PDF is the output, not the input.

**Reply by fmw42** [reply]
ImageMagick is a raster processor. The final PDF will be a raster image in a vector shell and not a vector image. Furthermore any resizing would involve resampling of your image or skipping of pixels or repeating of pixels depending upon how you resize it. ImageMagick does not do processing using PDF commands.

**Comment by snibgo** [comment]
> Ideally, the JPEG image is not resampled, just stretched using PDF commands.

IM can't generate whatever PDF commands (ie the code within a PDF file) would stretch an image.

**Comment by feklee** [comment]
Understood. *How would I do the stretching to image size if resampling is OK for me?*

I tried (A4 / 300 dpi):

    magick convert -page A4 -resize %[fx:page.width*300]x%[fx:page.height*300]! in.jpg out.pdf

However, that gives me the error message:

    convert: invalid argument for option '-resize': %[fx:page.width*300]x%[fx:page.height*300]! @ error/convert.c/ConvertImageCommand/2617.

It looks like `-resize` doesn't accept fx escapes. But maybe there is another simple solution. Of course I could do some computation outside of ImageMagick.

**Reply by fmw42** [reply]
For IM 7, use magick, not magick convert. The latter gives IM 6 behavior and would not accept the fx computations. But IM 7 with magick will work.

**Comment by myicq** [comment]
You can't specify page size "A4" without telling what PIXEL size you want (or reversed: what is your desired DPI). If you have 210 mm edge to edge, and want 300 Dots Per Inch, you have 11,811 dots per mm, and your (unscaled) image would have to be 2480 pixel wide. Any DPI different would change the resolution in PDF.

**Reply by feklee** [reply]
Right, `page.width` should be in pixels. In my code I falsely assume that it is in inches. Anyhow, the fx escapes don't seem to work with `-resize`. So I wonder if there is another way to do what I want purely with ImageMagick.

**Comment by feklee** [comment, marked as answer]
This works, for generating an A4 at 300dpi:

    magick in.jpg -density 72 -page A4 -resize %[fx:page.width*300/72]x%[fx:page.height*300/72]! out.pdf

Without specifying `-density 72`, the PDF is not 210x297 mm. Apparently, ImageMagick expresses dimensions of the hard coded page formats in pixels at 72 dpi.

With custom dimensions, here again A4 expressed in mm, it's possible to specify 300 dpi:

    magick in.jpg -density 300 -page %[fx:210/25.4*300]x%[fx:297/25.4*300] -resize %[fx:page.width]x%[fx:page.height]! out.pdf

**Reply by fmw42** [reply]
PDF output density is 72 dpi and I do not believe you can change that from my experience. So make your page size so that 72 dpi will print to the size you want.

**Reply by feklee** [reply]
Both of my examples have the desired output resolution of 300 DPI. So all is good.
